remove backend keys from FunctionalTensorWrapper, update TensorImpl::is_device methods

bdhirsh opened the pull request:
It's kinda annoying to have wrapper subclass tensors (like `FunctionalTensorWrapper` include backend dispatch keys in their keyset, because when we occasionally write something buggy, we'll send the wrapper tensor the the backend kernel (which usually segfaults). By ensuring that wrapper tensors don't get backend keys, we'll get a nicer error when that happens.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #81471

bdhirsh committed:
remove backend keys from FunctionalTensorWrapper, update TensorImpl::is_device methods

[ghstack-poisoned]

bdhirsh commented:
Since LTC/XLA are the only backends that have that special arrangement (of autograd directly seeing the `FunctionalTensorWrapper`), what do you think of compromising, and removing any backend bits from `FunctionalTensorWrapper` except for those two?

I'm sorta ok with it because criteria of "will autograd see a `FunctionalTensorWrapper` object is entirely backend dependent, and basically opted into from this code: https://github.com/pytorch/pytorch/blob/e3b98ba7a466ed98e3ca04714203ff7cb9d0dbe8/c10/core/DispatchKeySet.cpp#L27.

bdhirsh commented:
Oh okay, I have another reservation about making this change.

This PR will also prevent us from being able to make `Functionalize` a per-backend dispatch key; in order to do that, the tensors that dispatch to a functionalization kernel need to have backend bits in their keysets.

That conflicts with some of our potential future designs, around how decompositions should work longer term. Maybe we should hold off on this PR until that design is clearer?

ezyang commented on c10/core/TensorImpl.h:
```diff
@@ -837,8 +837,7 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
     }
-    constexpr auto meta_ks = DispatchKeySet(BackendComponent::MetaBit);
-    return key_set_.has_all(meta_ks);
+    return device_opt_.has_value() && device_opt_->type() == kMeta;
```
one to two memory derefs, but they should be in cache. Looks ok. I don't think these are called in tight loop anyway

bdhirsh commented:
> Here's an alternative: what if remove Dense only from the dispatch key set?

oh, that is a great idea :)

bdhirsh committed:
Update on "remove backend keys from FunctionalTensorWrapper, update TensorImpl::is_device methods"

It's kinda annoying to have wrapper subclass tensors (like `FunctionalTensorWrapper` include backend dispatch keys in their keyset, because when we occasionally write something buggy, we'll send the wrapper tensor the the backend kernel (which usually segfaults). By ensuring that wrapper tensors don't get backend keys, we'll get a nicer error when that happens.




[ghstack-poisoned]

bdhirsh committed:
Update on "remove backend keys from FunctionalTensorWrapper, update TensorImpl::is_device methods"

It's kinda annoying to have wrapper subclass tensors (like `FunctionalTensorWrapper` include backend dispatch keys in their keyset, because when we occasionally write something buggy, we'll send the wrapper tensor the the backend kernel (which usually segfaults). By ensuring that wrapper tensors don't get backend keys, we'll get a nicer error when that happens.




[ghstack-poisoned]

bdhirsh commented:
@pytorchbot rebase

bdhirsh committed:
Update on "remove backend keys from FunctionalTensorWrapper, update TensorImpl::is_device methods"

It's kinda annoying to have wrapper subclass tensors (like `FunctionalTensorWrapper` include backend dispatch keys in their keyset, because when we occasionally write something buggy, we'll send the wrapper tensor the the backend kernel (which usually segfaults). By ensuring that wrapper tensors don't get backend keys, we'll get a nicer error when that happens.




[ghstack-poisoned]

bdhirsh committed:
Update on "remove backend keys from FunctionalTensorWrapper, update TensorImpl::is_device methods"

It's kinda annoying to have wrapper subclass tensors (like `FunctionalTensorWrapper` include backend dispatch keys in their keyset, because when we occasionally write something buggy, we'll send the wrapper tensor the the backend kernel (which usually segfaults). By ensuring that wrapper tensors don't get backend keys, we'll get a nicer error when that happens.




[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge

ezyang commented on aten/src/ATen/FunctionalTensorWrapper.cpp:
```diff
@@ -30,6 +30,19 @@ void FunctionalTensorWrapper::set_constructor_metadata() {
+  // we need to know if we're dispatching to AutogradCPU or AutogradXLA).
+  // Instead, it's sufficient to remove the `Dense` dispatch key,
+  // which prevents us from accidentally trying to directly run a CPU/CUDA kernel.
+  key_set_ = key_set_.remove(c10::DispatchKey::Dense);
```
fwiw, if we ever get quantized or sparse functionalization, this wouldn't be quite right ;)

bdhirsh replied:
good point

ezyang commented on test/test_functionalization.py.bak:
```diff
@@ -0,0 +1,855 @@
+# Owner(s): ["module: codegen"]
```
delete this

ezyang reviewed (approved):
giving you the stamp but you've got an errant file in the PR rn

bdhirsh committed:
Update on "remove backend keys from FunctionalTensorWrapper, update TensorImpl::is_device methods"

It's kinda annoying to have wrapper subclass tensors (like `FunctionalTensorWrapper` include backend dispatch keys in their keyset, because when we occasionally write something buggy, we'll send the wrapper tensor the the backend kernel (which usually segfaults). By ensuring that wrapper tensors don't get backend keys, we'll get a nicer error when that happens.




[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge
